motionplan/cBiRRT.go

```diff
@@ -461,59 +458,69 @@ func (mp *cBiRRTMotionPlanner) smoothPath(
 	schan := make(chan node, 1)
 	defer close(schan)
 
-	for iter := 0; iter < toIter && len(inputSteps) > 4; iter++ {
-		select {
-		case <-ctx.Done():
-			mp.logger.Debug("CBiRRT timed out during smoothing, returning best path")
-			return inputSteps
-		default:
-		}
-		// Pick two random non-adjacent indices, excepting the ends
-
-		j := 2 + mp.randseed.Intn(len(inputSteps)-3)
-
-		i := mp.randseed.Intn(j) + 1
-
-		ok, hitCorners := smoothable(inputSteps, i, j, mp.corners)
-		if !ok {
-			continue
-		}
-
-		shortcutGoal := make(map[node]node)
-
-		iSol := inputSteps[i]
-		jSol := inputSteps[j]
-		shortcutGoal[jSol] = nil
-
-		// extend backwards for convenience later. Should work equally well in both directions
-		mp.constrainedExtend(ctx, mp.randseed, shortcutGoal, jSol, iSol, schan)
-		reached := <-schan
-
-		// Note this could technically replace paths with "longer" paths i.e. with more waypoints.
-		// However, smoothed paths are invariably more intuitive and smooth, and lend themselves to future shortening,
-		// so we allow elongation here.
-		dist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: inputSteps[i].Q(), EndConfiguration: reached.Q()})
-		if dist < mp.algOpts.JointSolveDist && len(reached.Q()) < j-i {
-			mp.corners[iSol] = true
-			mp.corners[jSol] = true
-			for _, hitCorner := range hitCorners {
-				mp.corners[hitCorner] = false
+	for numCornersToPass := 2; numCornersToPass > 0; numCornersToPass-- {
+		for iter := 0; iter < toIter/2 && len(inputSteps) > 4; iter++ {
```

Comment:
if instead of picking a random i, we just have i traverse the nodes of inputSteps sequentially, then I think we could potentially exit if no changes were made to the path on that iteration?

Reply:
> instead of picking a random i, we just have i traverse the nodes of inputSteps sequentially

This isn't desirable, because a smooth of a later i may enable/improve a smooth of an earlier i.

> the case where we have an input of length 4 and we repeatedly try the same thing over and over again

We probably want to reduce `numCornersToPass` to 1 in this case; probably we should have a function to determine the appropriate number of corners to pass based on path length.

Reply:
>We probably want to reduce numCornersToPass to 1 in this case; probably we should have a function to determine the appropriate number of corners to pass based on path length.

This would potentially work but we would have to add a heuristic and I'm not convinced that we gain much from this speedup.  I think we should leave as is

Reply:
It could be a very simple heuristic; especially since each successful smooth will add additional corners.

Something as simple as, if total number of corners is less than 7, only pass one corner, otherwise pass two.